File: algorithms/math/pollards-rho/python/pollards_rho.py
import math


def is_prime(n):
    if n < 2:
        return False
    if n < 4:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True


def rho(n):
    if n % 2 == 0:
        return 2
    x = 2
    y = 2
    c = 1
    d = 1
    while d == 1:
        x = (x * x + c) % n
        y = (y * y + c) % n
        y = (y * y + c) % n
        d = math.gcd(abs(x - y), n)
    if d != n:
        return d
    return n


def smallest_prime_factor(n):
    if n <= 1:
        return n
    if is_prime(n):
        return n
    factors = []
    stack = [n]
    while stack:
        num = stack.pop()
        if num == 1:
            continue
        if is_prime(num):
            factors.append(num)
            continue
        d = rho(num)
        if d == num:
            # Try different starting values
            for c in range(2, 20):
                x = 2
                y = 2
                d = 1
                while d == 1:
                    x = (x * x + c) % num
                    y = (y * y + c) % num
                    y = (y * y + c) % num
                    d = math.gcd(abs(x - y), num)
                if d != num:
                    break
        stack.append(d)
        stack.append(num // d)
    return min(factors) if factors else n


def pollards_rho(n):
    return smallest_prime_factor(n)


if __name__ == "__main__":
    print(pollards_rho(15))
    print(pollards_rho(13))
    print(pollards_rho(91))
    print(pollards_rho(221))

File: algorithms/math/pollards-rho/go/pollards_rho.go
package main

import (
	"fmt"
	"math/big"
)

func pollardsRho(n int64) int64 {
	if n <= 1 {
		return n
	}
	bn := big.NewInt(n)
	if bn.ProbablyPrime(20) {
		return n
	}
	// Trial division for small factors
	for p := int64(2); p*p <= n && p < 1000; p++ {
		if n%p == 0 {
			return p
		}
	}
	// Pollard's rho
	smallest := n
	stack := []int64{n}
	for len(stack) > 0 {
		num := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if num <= 1 {
			continue
		}
		bnum := big.NewInt(num)
		if bnum.ProbablyPrime(20) {
			if num < smallest {
				smallest = num
			}
			continue
		}
		d := rhoFactor(num)
		stack = append(stack, d, num/d)
	}
	return smallest
}

func rhoFactor(n int64) int64 {
	if n%2 == 0 {
		return 2
	}
	x, y, c, d := int64(2), int64(2), int64(1), int64(1)
	for d == 1 {
		x = mulmod(x, x, n)
		x = (x + c) % n
		y = mulmod(y, y, n)
		y = (y + c) % n
		y = mulmod(y, y, n)
		y = (y + c) % n
		diff := x - y
		if diff < 0 {
			diff = -diff
		}
		d = gcd64(diff, n)
	}
	if d != n {
		return d
	}
	return n
}

func mulmod(a, b, m int64) int64 {
	ba := big.NewInt(a)
	bb := big.NewInt(b)
	bm := big.NewInt(m)
	ba.Mul(ba, bb)
	ba.Mod(ba, bm)
	return ba.Int64()
}

func gcd64(a, b int64) int64 {
	for b != 0 {
		a, b = b, a%b
	}
	if a < 0 {
		return -a
	}
	return a
}

func main() {
	fmt.Println(pollardsRho(15))
	fmt.Println(pollardsRho(13))
	fmt.Println(pollardsRho(91))
	fmt.Println(pollardsRho(221))
}
